listenerutil/handler.go

```diff
@@ -0,0 +1,101 @@
+package listenerutil
+
+import (
+	"net/http"
+)
+
+type ResponseWriter struct {
+	wrapped http.ResponseWriter
+	// headers contain a map of response code to header map such that
+	// headers[status][header name] = header value
+	// this map also contains values for hundred-level values in the format 1: "1xx", 2: "2xx", etc
+	// defaults are set to 0
+	headers       map[int]map[string]string
+	headerWritten bool
+}
+
+func (w *ResponseWriter) WriteHeader(statusCode int) {
+	w.headerWritten = true
+	w.setCustomResponseHeaders(statusCode)
+	w.wrapped.WriteHeader(statusCode)
+}
+
+func (w *ResponseWriter) Header() http.Header {
+	return w.wrapped.Header()
+}
+
+func (w *ResponseWriter) Write(data []byte) (int, error) {
+	// The default behavior of http.ResponseWriter.Write is such that if WriteHeader has not
+	// yet been called, it calls it with the below line. We will copy that logic so that our
+	// WriteHeader function is called rather than http.ResponseWriter.WriteHeader
+	if !w.headerWritten {
+		w.WriteHeader(http.StatusOK)
+	}
+	return w.wrapped.Write(data)
+}
+
+func (w *ResponseWriter) setCustomResponseHeaders(statusCode int) {
+	sch := w.headers
+	if sch == nil {
+		return
+	}
+
+	// Check the validity of the status code
+	if statusCode >= 600 || statusCode < 100 {
+		return
+	}
+
+	// Setter function to set headers
+	setter := func(headerMap map[string]string) {
+		for header, value := range headerMap {
+			w.Header().Set(header, value)
+		}
+	}
+
+	// Setting the default headers first
+	if val, ok := sch[0]; ok {
+		setter(val)
+	}
+
+	// Then setting the generic hundred-level headers
+	if val, ok := sch[statusCode/100]; ok {
+		setter(val)
+	}
+
+	// Finally setting the status-specific headers
+	if val, ok := sch[statusCode]; ok {
+		setter(val)
+	}
+}
+
+type uiRequestFunc func(*http.Request) bool
+
+// WrapCustomHeadersHandler wraps the handler to pass a custom ResponseWriter struct to all
+// later wrappers and handlers to assign custom headers by status code. This wrapper must
+// be the outermost wrapper to function correctly.
```

Comment:
The terminology for this is a HTTP middleware, lets use that in the API docs. Also, this describes the internal implementations details, when as a user I just want to know that it does what I want with the config.
```suggestion
// WrapCustomHeadersHandler wraps the handler, adding custom headers as defined in the
// ListenerConfig. The isUiRequest callback is used to determine whether an incoming request
// should get the UI specific headers or the generic API headers. This middleware must
// be the outermost middleware to function correctly in the case of another middleware returning
// early.
```

Reply:
Actually that reminds me, what happens if this middleware wraps a handler that returns without calling w.WriteHeader? Should we call w.WriteHeader for good measure before returning from here?

Reply:
I have another comment addressing this in the code [here](https://github.com/hashicorp/go-secure-stdlib/pull/49/files#diff-edb1c099ae693e6b00aa24b159e9cd9cb6aca1bab17f5d168bf0ae5b90757138R28), but the default behavior of `ResponseWriter` is such that if a call has not been made to `WriteHeader` when `Write` is called, then `WriteHeader(200)` is called automatically

Reply:
Yes, but that will happen outside of our custom response writer, will it not?

Reply:
if you check out the link I posted, you can see the logic is replicated in the response writer and is even tested for in the handler tests for the 200 responses

Reply:
My concern is what happens if `Write` isn't called but the inner middleware simply returns? It'd get all the way out of the middleware chain and then call `WriteHeader` on a response writer that isn't ours, no?